src/lib.rs

```diff
@@ -17,8 +20,95 @@ use tempfile::TempDir;
 use tracing::{debug, info, instrument, warn};
 use url::Url;
 
+use hc_utils::WrappedHeaderHash;
+use holochain::conductor::api::AppResponse;
+use holochain::conductor::api::ZomeCall;
+use holochain_zome_types::{
+    zome::{FunctionName, ZomeName},
+    zome_io::ExternInput,
+    SerializedBytes,
+};
 type HappIds = Vec<String>;
 
+pub async fn activate_holo_hosted_happs(core_happ: Happ) -> Result<()> {
+    let list_of_happs: Vec<WrappedHeaderHash> = get_enabled_hosted_happs(core_happ).await?;
+    info!("got back list_of_happs {:?}", list_of_happs);
+    install_holo_hosted_happs(list_of_happs).await?;
+    Ok(())
+}
+
+pub async fn install_holo_hosted_happs(happs: Vec<WrappedHeaderHash>) -> Result<()> {
+    info!("Starting to install....");
+    // iterate through the vec and
+    // Call http://localhost/hpos-holochain-api/install_hosted_happ
+    // for each WrappedHeaderHash to install the hosted_happ
+    let client = reqwest::Client::new();
+    // Note: Tmp preferences
+    let preferences = Preferences {
+        max_fuel_before_invoice: 1.0,
+        max_time_before_invoice: vec![86400, 0],
+        price_compute: 1.0,
+        price_storage: 1.0,
+        price_bandwidth: 1.0,
+    };
+    for happ_id in happs {
+        info!("Installing happ-id {:?}", happ_id);
+        let body = Body {
+            happ_id: happ_id.0.to_string(),
+            preferences: preferences.clone(),
+        };
+        let response = client
+            .post("http://localhost/hpos-holochain-api/install_hosted_happ")
+            .json(&body)
+            .send()
+            .await?;
+        info!("Installed happ-id {:?}", happ_id);
+        info!("Response {:?}", response);
+    }
+    Ok(())
+}
+
+pub async fn get_enabled_hosted_happs(core_happ: Happ) -> Result<Vec<WrappedHeaderHash>> {
+    let mut app_websocket = AppWebsocket::connect(42233)
+        .await
+        .context("failed to connect to holochain's app interface")?;
+    match app_websocket.get_app_info(core_happ.id_from_config()).await {
+        Some(app_info) => {
+            let zome_call_payload = ZomeCall {
+                cell_id: app_info.cell_data[0].clone().into_id(), // This works on the assumption that the core happs has HHA in the first position of the vec
+                zome_name: ZomeName::from("hha"),
+                fn_name: FunctionName::from("get_happs"),
+                payload: ExternInput::new(SerializedBytes::default()),
+                cap: None,
+                provenance: app_info.cell_data[0]
+                    .clone()
+                    .into_id()
+                    .agent_pubkey()
+                    .to_owned(),
+            };
```

Comment:
this business with cloning provenance two times looks ugly/inefficient. Here's a way of rewriting this to take ownership of the necessary values and avoid cloning
```suggestion
        Some(holochain_types::InstalledApp {
            // This works on the assumption that the core happs has HHA in the first position of the vec
            cell_data: [hha_cell, ..],
            ..
        }) => {
            let zome_call_payload = ZomeCall {
                cell_id: hha_cell.as_id().clone(),
                zome_name: ZomeName::from("hha"),
                fn_name: FunctionName::from("get_happs"),
                payload: ExternInput::new(SerializedBytes::default()),
                cap: None,
                provenance: hha_cell
                    .into_id()
                    .into_dna_and_agent()
                    .1
            };
```

Reply:
It looks like `Vec` cannot be de-structured like `[happ_cell, ...]`.

Reply:
Darn it!! I'll make a new suggestion using `into_iter()` to take ownership then